rustls/src/client/tls13.rs

```diff
@@ -396,8 +396,17 @@ impl State<ClientConnectionData> for ExpectEncryptedExtensions {
 
         #[cfg(feature = "quic")]
         {
-            // QUIC transport parameters
             if cx.common.is_quic() {
+                // RFC 9001 says: While ALPN only specifies that servers use this alert, QUIC
+                // clients MUST use error 0x0178 to terminate a connection when ALPN negotiation
+                // fails.
+                if cx.common.alpn_protocol.is_none() && !self.config.alpn_protocols.is_empty() {
```

Comment:
So what this leaves implicit, I think (here and on the server side), is the spec language around if/how ALPN is mandatory. In 8.1, it says "Unless another mechanism is used for agreeing on an application protocol, endpoints MUST use ALPN for this purpose."; I think you're translating that into the `!self.config.alpn_protocols.is_empty()` here, but it would be nice to make that more explicit.

Also, have you considered putting this in `process_alpn_protocol()`, separately guarded by `cx.common.is_quic()`? While that function is intended for both 1.2 and 1.3, it feels like this logic does fit in with what you've added here.

Reply:
Moved the logic into `process_alpn_protocol` (where it indeed feels natural) and expanded on the rationale a bit.